### PR TITLE
feat(local): add logic for handling stages

### PR DESCRIPTION
**Pull request by jbrockopp**
Part of the effort for https://github.com/go-vela/community/issues/54

This implements a set of functions for dealing with `stages` in the `local` executor client.

* `CreateStage()`
* `PlanStage()`
* `ExecStage()`
* `DestroyStage()`

This is a stepping stone to getting the `local` executor fully implemented when trying to run a pipeline locally.

### CODE REVIEW

**Review by kneal — APPROVED** [review]
LGTM 🐬

File: executor/local/stage.go [hunk 1]
```diff
@@ -6,26 +6,169 @@ package local
...
+	// ensure dependent stages have completed
+	for _, needs := range s.Needs {
+		// check if a dependency stage has completed
+		stageErr, ok := m[needs]
+		if !ok { // stage not found so we continue
+			continue
+		}
+
+		// wait for the stage channel to close
+		select {
```
Comment (matt-fevold) [inline executor/local/stage.go:70]: I hadn't seen this before! [It](https://gobyexample.com/select) seems pretty powerful combined with [receivers](https://golang.org/ref/spec#Receive_operator).

File: executor/local/stage.go [hunk 2]
```diff
@@ -6,26 +6,169 @@ package local
...
+	}
+
+	// update the init log with progress
+	//
+	// https://pkg.go.dev/github.com/go-vela/types/library?tab=doc#Log.AppendData
+	l.AppendData([]byte(fmt.Sprintf("> Pulling step images for stage %s...\n", s.Name)))
+
+	// create the steps for the stage
+	for _, step := range s.Steps {
+		// TODO: make this not hardcoded
```
Comment (matt-fevold) [inline executor/local/stage.go:32]: Assuming this todo will be updated in a later PR?

**Review by matt-fevold — APPROVED** [review]
LGTM

File: executor/local/stage.go (hunk 1 above)
Comment (jbrockopp) [reply executor/local/stage.go:70]: Yes you are correct 👍 

Typically I've only seen it used when trying to handle "infinite" loops or concurrency primitives in Go.

But it sure does come in handy when you need it 😅

File: executor/local/stage.go (hunk 2 above)
Comment (jbrockopp) [reply executor/local/stage.go:32]: Yes you are correct 👍 

I plan to refactor chunks of the codebase once I have a functional version of the `local` executor

**Pull request merged**